### ISSUE TITLE
[UX] Move the 403/404 path settings to a better suited place.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#4457]
It feels that these settings have been "thrown" in that page for lack of a better place...

I feel that somewhere under "URL handling" may be a better place.

---
PR that moves error settings into `URL redirect` section: https://github.com/backdrop/backdrop/pull/3187
PR that moves error settings into `Error handling` section: https://github.com/backdrop/backdrop/pull/3197

(Triaged as possible accessibility issue on 08/17/2921. This one only came up because a possible solution might be bad for accessibility, please add the [A11Y] tag if you think it qualifies.)

**Comment by jenlampton** [comment]
I thought I created an issue for this way back...  We already have a `Logging and errors` section, and since 403 and 404 are both errors, I put them there.

**Comment by klonos** [comment]
Hmm, I don't think that under Development -> Logging and errors is a good fit for these settings 🤔 ...that place is for php errors and the likes, and for logging these errors. I was thinking either under Configuration -> URL handling -> URL settings (`admin/config/urls/settings`):

![image](https://user-images.githubusercontent.com/2423362/87322273-e094ad00-c535-11ea-9e35-f8b017e6bd34.png)


...or under Configuration -> URL handling -> URL redirects -> Settings (`admin/config/urls/redirect/settings`):

![image](https://user-images.githubusercontent.com/2423362/87322472-1fc2fe00-c536-11ea-9eb9-99f3e501b8fc.png)

...the fact that we have the "Fix 404 pages" tab there makes me think that it's even more fitting.

**Comment by klonos** [comment]
...perhaps it should even be under that "Fix 404 pages" tab somewhere 🤷

**Comment by klonos** [comment]
I took a shot at this, moving the settings under `admin/config/urls/redirect`. Here's what it looks like:

![Screen Shot 2020-07-14 at 1 13 54 am](https://user-images.githubusercontent.com/2423362/87358930-7c8eda80-c56f-11ea-80a5-87241e6bd9dc.png)

- The "Settings" tab has been renamed to "Redirect settings", and placed right after the "List URL redirects" tab
- A new "4xx error settings" tab has been introduced, to hold the 404/403 settings form.
- The "4xx error settings" tab is placed after the "Fix 404 pages" tab.
- Since I do not like (UX-wise) the Drupalism that is the "leave blank to ..." help text, I've replaced that with checkboxes.

We have a few options here:

1. Move the "Fix 404 pages" tab/menu one level up under the "URL handling" menu, and then call the "4xx error settings" tab just "Settings" (then also rename the "Redirects settings" tab to just "Settings").
2. Make the "Redirects settings" be a secondary tab, under "List URL redirects" (and just call it "Settings") + Make the "4xx error settings" be a secondary tab, under "Fix 404 pages" (and just call it "Settings").
3. The way it is now, the "Display a generic 404/403 page" checkboxes behave the same as the "Use the logo supplied by the active theme" checkbox: when ticked, the respective option to specify a 404/403 custom path is hidden. We could change them to be "Use a custom 404/403 page" instead, and have them behave the opposite: the option to specify a 404/403 custom path is hidden by default, and only shown if the respective checkbox is ticked. It would look like this:

![Screen Shot 2020-07-14 at 2 50 06 am](https://user-images.githubusercontent.com/2423362/87364760-245ed500-c57d-11ea-98cc-f0abf73aae2b.png)

**Comment by jenlampton** [comment]
no no I really don't like having important site settings buried in the unrelated "URL Redirects" section :(

**Comment by jenlampton** [comment]
> Hmm, I don't think that under `Development -> Logging and errors` is a good fit for these settings 🤔 ...that place is for php errors and the likes, and for logging these errors.

Is a difference? Both 404 errors and 503 errors are logged. Would it make you feel better if we changed the label to `Errors and Logging` ?

**Comment by klonos** [comment]
> So, why are we thinking about moving this setting in with URL redirects?

That wasn't the goal. I was thinking to put them next/close to the "Fix 404 pages" page.

**Comment by jenlampton** [comment]
> put them next/close to the "Fix 404 pages" page.

Yes, the "Fix 404 pages" page is a page where you create URL redirects -- that's why it's in the `URL Redirects` section.

**Comment by klonos** [comment]
Yeah, I was thinking that in the 404/403 admin pages we do not specify where/how these requests are being logged, rather where we "redirect" people once that happens.

**Comment by jenlampton** [comment]
But we aren't redirecting anyone anywhere. This setting as the same as the "home page" setting we used to have be `/node` but is now `/home`. URL redirects are a specific thing, and they can all be viewed in this list `admin/config/urls/redirect`. The two concepts are unrelated.

**Comment by jenlampton** [comment]
@klonos thanks for the Jitsi chat, I see your train of thought now. I still prefer the error display and logging settings location than the URL redirects lcoation, but I agree neither are perfect. I don't know if these two settings are enough to justify their own page, but that's how they end up on the site information page :) 

Another thought for discussion: what do we even call these things? Are they `Replacement pages` maybe?

**Comment by indigoxela** [comment]
Wherever you think the best place might be: keep in mind to not put it somewhere under a (possibly disabled) module setting.
The custom error pages setting also applies even if module redirect or path or dblog aren't installed.

(Personally I'm fine with the current place.)

**Comment by klonos** [comment]
> ...keep in mind to not put it somewhere under a (possibly disabled) module setting.

Great catch there @indigoxela 👍

**Comment by klonos** [comment]
> (Personally I'm fine with the current place.)

During our chat with @jenlampton we've hit thought train dead-ends multiple times, and we've both agreed that this whole discussion seemed an awful lot like a déjà vu. Perhaps we've discussed it in the past, and that's why these things ended up where they currently are 😅 🤷 

We both dislike the fact that Configuration -> System seems to be treated like a "dumping ground" for things that do not fit anywhere else, but we couldn't find a better-suited place. From a non-Drupal/Backdrop point of view, this does not make sense, and I don't think that people would be able to discover these settings if they were looking for them ...only perhaps by accident, after pogo-sticking in the admin UI.

We did also look into a recent Wordpress demo (v5.4.2), and they are not doing any better a job either 😞 ...in their case, configuring their 404 page requires editing code (via the UI 🤦 🙄 ). This is placed under Appearance -> Theme editor and then relying on the current theme providing a "404 Template" to edit. If the theme does not provide one, you'll need to copy `/wp-content/themes/twentythirteen/404.php` into the directory of your current theme ("Twenty Thirteen" is the default core theme in WP).

Feedback/thoughts welcome.

**Comment by indigoxela** [comment]
Maybe you're hunting ghosts? :wink:

The setting for custom 40x error pages is a system wide setting, it's independent from other modules' functionality. So to me it actually appears at the best available place.

I'm a bit astounded that this setting seems to cause some head scratching. Maybe the problem is rather a wording thing with the admin menu item: "Site information" (vs. something like "Site wide settings" or "Global settings" or "Base settings" or... whatever's more appropriate)

**Comment by klonos** [comment]
Sure @indigoxela, when wearing a Drupal/Backdrop/developer hat, this all makes sense; but from an end user point of view, would the average site builder be able to find these settings easily, without pogo-sticking around the admin UI?

**Comment by jorgemare** [comment]
Howdy,

First time I post here. Intrigued by Backdrop, have been following the project on the side for a while, and like your user-centric approach. I have not used Backdrop yet, though, so I am not familiar with the system, so forgive me if I am off with my observation.

I feel it would be more natural (intuitivel) to have an Error Handling page with subsections (tabs?) for settings like Error Display, DB log settings, Syslog settings, and 403/404 pages.

Food for thought.

**Comment by jenlampton** [comment]
> "Site information" (vs. something like "Site wide settings" or "Global settings" or "Base settings" or... whatever's more appropriate)

This is a good point. 

We're trying to turn the `Site information` page into a page that contains only Site information -- but since this page is currently a dumping ground for more general settings, maybe the easier solution would be to change the name of the page to `General Settings`, with a section for `Site information`?

> an Error Handling page with subsections (tabs?) for settings like Error Display, DB log settings, Syslog settings, and 403/404 pages.

This is what I was going for with my recommendation too. By moving the error handling settings over to an `Error handling` section, we also turn the `Site information` section back into a page with only site information. 

I think @klonos also had up a good point that people might not see `Logging and errors` as an `Error handling` section right now. People might think of it as a `Logging section` instead. 

Maybe if we changed the label from `Logging and errors` to `Error handling` and added fieldsets to that page as part of the patch, that might help?  I've created an alternate PR for comparison: https://github.com/backdrop/backdrop/pull/3197

**Comment by klonos** [comment]
Thanks @jorgemare and @indigoxela for chiming in with really helpful observations and suggestions, and thank you @jenlampton for the PR; things are starting to make a lot more sense now 👍  ...I've closed my PR in favor of yours, and left a couple of minor suggestions, but this looks and works nicely in general; just need to fix the failing tests.

**Comment by klonos** [comment]
...I forgot to mention that I'm still annoyed by the "Leave blank to display a generic ... page." messages, which is a nasty Drupalism. In my PR, I converted those to checkboxes, same as we do for the logo and favicon. We can do that here if you think that it's a UX improvement, or it can be a follow-up.

**Comment by jenlampton** [comment]
If you already have the code, would you please file another PR against mine?  :)

**Comment by klonos** [comment]
Done @jenlampton: https://github.com/jenlampton/backdrop/pull/13

**Comment by jorgemare** [comment]
Howdy,

Saw the Error Handling page on the test site, and looks good! I have another observation.

When I started using Drupal, the Error Messages to Display setting confused me. It made me think that if I chose "None", then Drupal would display no error messages at all. It took me some time to figure out that that was not the case.

To avoid this potential confusion among new users, you may want to consider somehow distinguishing between server/PHP messages and CMS messages.

Other than that, looks like the right direction IMHO!

**Comment by klonos** [comment]
> ...you may want to consider somehow distinguishing between server/PHP messages and CMS messages.

Good point @jorgemare 👍

**Comment by jenlampton** [comment]
@klonos I used most of your PR as it was, but I was struggling to find good places for all the help text. before all the changes, the description text read `This page is displayed when the requested document is denied to the current user. Leave blank to display a generic "access denied" page.`  That message did two things: 1) it explained what the page was and when it would appear (help text) and 2) it told the user what would happen if a custom error page were not defined. 

It occurred to me that even when adding a checkbox to enable/disable the custom path setting, we still need to include this second bit of information, but there wasn't a good place for it in your PR (which I expect, was why it was excluded).

I've tried switching to radios to properly convey the two options people have for these pages. Let me know what you think?

<img width="1196" alt="Screen Shot 2020-07-27 at 10 19 39 AM" src="https://user-images.githubusercontent.com/397895/88572921-d7571600-cff4-11ea-936b-89c8e7b4a2d8.png">

**Comment by jenlampton** [comment]
from a DX perspective I wanted to change the ID of the form from `system_logging_settings` to `system_error_settings` but when I started to do that, I realized that both `dblog` and `syslog` alter that form using the `form_id`. Though this too would be easy enough to change in core, I stopped when I wondered if other contrib modules might do the same. 

I have a separate PR with these changes here: https://github.com/backdrop/backdrop/pull/3202

**Comment by klonos** [comment]
> ...but there wasn't a good place for it in your PR (which I expect, was why it was excluded).

Oh, I simply missed that 😅 ...we could add a single help/description in the fieldset for both the custom 404/403 fields, which IMO would keep the complexity/noise on the form down:

![image](https://user-images.githubusercontent.com/2423362/88599992-d914ee00-d075-11ea-9460-2ddd57c51323.png)

**Comment by klonos** [comment]
...I liked the wording changes in the help text for both fields @jenlampton 👍

**Comment by jenlampton** [comment]
> we could add a single help/description

Yeah, I ended up liking the description option better too, and I did that in the current PR: 

<img width="1209" alt="Screen Shot 2020-07-27 at 4 31 21 PM" src="https://user-images.githubusercontent.com/397895/88602009-abed1f00-d026-11ea-8453-f6caeaba7781.png">

**Comment by klonos** [comment]
> I realized that both `dblog` and `syslog` alter that form using the `form_id` ... I stopped when I wondered if other contrib modules might do the same. 

I imagine that we will come across this again in the future. Would it be acceptable to change backdrop_prepare_form() so that it calls a new helper function similar to `backdrop_merged_modules()`, like this:

```php
function backdrop_prepare_form($form_id, &$form, &$form_state) {
  global $user;

  $form_id = backdrop_deprecated_form_ids($form_id);

  ...
}

...

function backdrop_deprecated_form_ids($form_id = NULL) {
  // Map old/deprecated form IDs to their respective new IDs.
  $deprecated_form_ids = array(
    'system_logging_settings' => 'system_error_settings', // Backdrop 1.17.0.
    ...
  );

  // Allow calling this function w/o any specific form ID, in which
  // case the entire array of deprecated form IDs is returned.
  if ($form_id == NULL) {
    return $deprecated_form_ids;
  }
  // If a specific $form_id was passed in the function, and it is
  // found among the deprecated ones, then return its replacement.
  elseif (in_array($form_id, $deprecated_form_ids)) {
    return $deprecated_form_ids[$form_id];
  }

  // If the specified $form_id was not found in the list of
  // deprecated ones, then return the same.
  return $form_id;
}
```

Thoughts?

**Comment by klonos** [comment]
>Yeah, I ended up liking the description option better too, and I did that in the current PR:

@jenlampton either I don't get it, or you posted the wrong screenshot 😅

**Comment by jenlampton** [comment]
The screenshot shows description text on the radios, instead of separate
help elements.

**Comment by klonos** [comment]
Right! ...I thought you meant a single description/help text added to the fieldset. Sorry.

With radios instead of checkboxes, and with all the field labels visible, I find that part of the form too "noisy"; so I prefer single checkboxes, and hiding some of the field labels (where we simply mostly duplicate wording). I'll let others chime in to say what they think, but I wouldn't wanna derail/delay the main goal of this issue; so if there is no consensus, we can split this into a follow-up.

**Comment by jenlampton** [comment]
> With radios instead of checkboxes, and with all the field labels visible, I find that part of the form too "noisy"; so I prefer single checkboxes,

Rather than removing important information (what will happen if you don't check the box) I think we should work on reducing the verbosity of all the labels and descriptions. 

> and hiding some of the field labels

We should never hide a field label because that's not good for accessibility. The label needs to go with the element so that everyone knows what goes in the element.

> if there is no consensus, we can split this into a follow-up.

agreed :)

**Comment by klonos** [comment]
> We should never hide a field label because that's not good for accessibility.

@jenlampton please see my comment in https://github.com/backdrop/backdrop/pull/3197#discussion_r460517847 re using `'#title_display' => 'invisible'`, which retains accessibility while at the same time visually hiding the label.
